inference-engine/src/multi_device/multi_device_exec_network.cpp

```diff
@@ -60,71 +90,221 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const DeviceMap<Infer
-                        Task t;
-                        if (_inferPipelineTasks.try_pop(t))
-                            ScheduleToWorkerInferRequest(std::move(t));
-                        else if (_inferPipelineTasksDeviceSpecific[device]->try_pop(t))
-                            ScheduleToWorkerInferRequest(std::move(t), device);
-                    }
-                });
+            _networkFirstReady = _cpuFuture.get();
+        } catch (const std::exception& e) {
+            printf("Warning: load network to CPU failed: %s\n", e.what());
+            _networkActualNeeded = _acceleratorFuture.get();
         }
+    } else if (_acceleratorFuture.valid()) {  // only accelerator is valid, like AUTO:GPU
+        _networkActualNeeded = _acceleratorFuture.get();
+    } else if (_cpuFuture.valid()) {  // only CPU is valid, like AUTO:CPU
+        _networkActualNeeded = _cpuFuture.get();
+    } else {
+        IE_THROW() << "No device task available";
+    }
+
+    // if there is only one device or loading CPU device is failed,
+    // the ActualNetwork is already ok now.
+    if (!_acceleratorFuture.valid()) {
+        _alreadyActualNetwork = true;
     }
 }
 
+void MultiDeviceExecutableNetwork::WaitActualNetworkReady() const {
+    // Maybe different API will call this function, so add call once here
+    // for every MultiDeviceExecutableNetwork instance
+    std::call_once(_oc, [&] () {
+            if (_acceleratorFuture.valid()) {
+                _networkActualNeeded = _acceleratorFuture.get();
+            }
+            });
+}
+
 void MultiDeviceExecutableNetwork::ScheduleToWorkerInferRequest(Task inferPipelineTask, DeviceName preferred_device) {
+    // AUTO work mode
+    if (_workModeIsAUTO) {
+        std::string selectDeviceName = "";
+        if (!preferred_device.empty()) {
+            // the preferred_device should be the selected device in AUTO work mode
+            if (preferred_device != _acceleratorDevice.deviceName) {
+                IE_THROW(NotFound) << "The preferred_device should be the selected device";
+            }
+            // if the device needed by customer is not ready, need to wait for it
+            WaitActualNetworkReady();
+            selectDeviceName = preferred_device;
+        } else {
+            // _acceleratorDevice could be the same as _cpuDevice, such as AUTO:CPU
+            selectDeviceName = _alreadyActualNetwork ? _acceleratorDevice.deviceName
+                                                     : _cpuDevice.deviceName;
+        }
+        if (selectDeviceName.empty()) {
+            IE_THROW(NotFound) << "The select device should not be empty";
+        }
+        if (RunPipelineTask(inferPipelineTask, _idleWorkerRequests[selectDeviceName], preferred_device)) {
+            return;
+        }
+        // no vacant requests this time, storing the task to the respective queue
```

Comment:
although minor, but the 3 lines below are the same on the AUTO and MULTI codepth in this function.
In fact the line 288 is ~same to the line 307 too:
(RunPipelineTask(inferPipelineTask, _idleWorkerRequests[selectDeviceName], preferred_device 

Again, ideally the only diff between AUTO and MULTI scheduling is selectDeviceName logic, the actual scheduling ( or putting the request back to the queue) is SAME

Reply:
shall we refactor this part to cleary show that only the names manipulations are different?